packages/fx-core/src/component/driver/deploy/azure/impl/AzureZipDeployDriverImpl.ts

```diff
@@ -0,0 +1,213 @@
+   * @param logger log provider
+   * @protected
+   */
+  protected async zipDeployPackage(
+    zipDeployEndpoint: string,
+    zipBuffer: Buffer,
+    config: AzureUploadConfig,
+    logger?: LogProvider
+  ): Promise<string> {
+    let res: AxiosZipDeployResult;
+    let retryCount = 0;
+    while (true) {
+      try {
+        res = await AzureDeployDriverImpl.AXIOS_INSTANCE.post(zipDeployEndpoint, zipBuffer, config);
+        break;
+      } catch (e) {
+        if (axios.isAxiosError(e)) {
+          // if the error is remote server error, retry
+          if ((e.response?.status ?? HttpStatusCode.OK) >= HttpStatusCode.INTERNAL_SERVER_ERROR) {
+            retryCount += 1;
+            if (retryCount < DeployConstant.DEPLOY_UPLOAD_RETRY_TIMES) {
+              await logger?.warning(
+                `Upload zip file failed with response status code: ${
+                  e.response?.status ?? "NA"
+                }. Retrying...`
+              );
+            } else {
+              // if retry times exceed, throw error
+              await logger?.warning(
+                `Retry times exceeded. Upload zip file failed with remote server error. Message: ${JSON.stringify(
+                  e.response?.data
+                )}`
+              );
+              throw DeployExternalApiCallError.zipDeployWithRemoteError(
+                e,
+                undefined,
+                this.helpLink
+              );
+            }
+          } else {
+            // None server error, throw
+            await logger?.error(
+              `Upload zip file failed with response status code: ${
+                e.response?.status ?? "NA"
+              }, message: ${JSON.stringify(e.response?.data)}`
+            );
+            throw DeployExternalApiCallError.zipDeployError(
+              e,
+              e.response?.status ?? -1,
+              this.helpLink
+            );
+          }
+        } else {
+          // if the error is not axios error, throw
+          await logger?.error(`Upload zip file failed with error: ${JSON.stringify(e)}`);
+          throw DeployExternalApiCallError.zipDeployError(e, -1, this.helpLink);
+        }
+      }
+    }
+
+    if (res?.status !== HttpStatusCode.OK && res?.status !== HttpStatusCode.ACCEPTED) {
+      if (res?.status) {
+        await logger?.error(`Deployment is failed with error code: ${res.status}.`);
+      }
+      throw DeployExternalApiCallError.zipDeployError(res, res.status, this.helpLink);
+    }
+
+    return res.headers.location;
+  }
+
+  /**
+   * create azure zip deploy endpoint
+   * @param siteName azure app service or azure function name
+   * @protected
+   */
+  protected getZipDeployEndpoint(siteName: string): string {
+    return `https://${siteName}.scm.azurewebsites.net/api/zipdeploy?isAsync=true`;
+  }
+
+  createProgressBar(ui?: UserInteraction): IProgressHandler | undefined {
+    const steps = this.dryRun ? 1 : 6;
```

Comment:
This looks error prone. I will forget to edit the number when I add a new progress step.

Reply:
Maybe we can make it as a constant to keep it more meaningful and remind us to update when step is added/removed. Or there is some way to auto calculated the steps by some definition/configuration?